### PR TITLE
Add `opaque_inspect_tensor` wrapper around @leaf_function for printing in compiled code

### PR DESCRIPTION
**Pull request by yushangdi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #175298
* __->__ #174636
* #174635

## Summary

Add `make_opaque_inspect_tensor_fn`, a utility for observing tensor values during forward and backward passes that works seamlessly across eager mode, `torch.compile`, and `make_fx`.

Built on top of `@leaf_function`, both the forward and backward callbacks are opaque to the compiler — they execute at runtime without being traced.

### API

```python
from torch._higher_order_ops.opaque_inspect_tensor import make_opaque_inspect_tensor_fn

opaque_inspect = make_opaque_inspect_tensor_fn(
    fwd_f=callable,  # called as fwd_f(tensor) on each tensor in forward
    bwd_f=callable,  # called as bwd_f(grad) on each gradient in backward
)

# Call inside compiled or eager code:
opaque_inspect(*tensors, tag="", ranks=None, phase=None)
```

**Parameters of the returned callable:**
- `tag` — optional label, prints `[tag][fwd]` / `[tag][bwd]` prefixes
- `ranks` — filter by distributed rank (`None` = all, `int` = single, `set[int]` = multiple)
- `phase` — override which callback runs:
  - `None` (default): calls `fwd_f`, registers backward hooks for `bwd_f`
  - `"fwd"`: calls `fwd_f` only, no backward hooks
  - `"bwd"`: calls `bwd_f` directly, no backward hooks

### Example — basic usage (works in both eager and compiled)

```python
opaque_inspect = make_opaque_inspect_tensor_fn(
    fwd_f=lambda t: print(f"  fwd: shape={t.shape}, norm={t.norm():.4f}"),
    bwd_f=lambda t: print(f"  bwd: shape={t.shape}, norm={t.norm():.4f}"),
)

def fn(x):
    opaque_inspect(x)
    return x.sum()

# Eager
x = torch.randn(3, 3, requires_grad=True)
fn(x).backward()
#   fwd: shape=torch.Size([3, 3]), norm=2.5700
#   bwd: shape=torch.Size([3, 3]), norm=3.0000

# Compiled — same function, same output
compiled_fn = torch.compile(fn, backend="aot_eager")
compiled_fn(x).backward()
#   fwd: shape=torch.Size([3, 3]), norm=2.5700
#   bwd: shape=torch.Size([3, 3]), norm=3.0000
```

### Example — tagging with multiple tensors

```python
opaque_inspect = make_opaque_inspect_tensor_fn(
    fwd_f=lambda t: print(f"    {t.shape}"),
    bwd_f=lambda t: print(f"    {t.shape}"),
)

x = torch.randn(2, 4, requires_grad=True)
y = torch.randn(2, 4, requires_grad=True)
opaque_inspect(x, y, tag="inputs")
# [inputs][fwd]
#     torch.Size([2, 4])
#     torch.Size([2, 4])
(x + y).sum().backward()
# [inputs][bwd]
#     torch.Size([2, 4])
# [inputs][bwd]
#     torch.Size([2, 4])
```

### Example — inside a custom autograd function

Use `phase="fwd"` / `phase="bwd"` to control which callback runs:

```python
from torch.autograd import Function

opaque_inspect = make_opaque_inspect_tensor_fn(
    fwd_f=lambda t: print(f"  shape={t.shape}, norm={t.norm():.4f}"),
    bwd_f=lambda t: print(f"  shape={t.shape}, norm={t.norm():.4f}"),
)

class MyReLU(Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        opaque_inspect(x, tag="my_relu", phase="fwd")    # prints [my_relu][fwd]
        return x.clamp(min=0)

    @staticmethod
    def backward(ctx, grad_output):
        (x,) = ctx.saved_tensors
        opaque_inspect(grad_output, tag="my_relu", phase="bwd")  # prints [my_relu][bwd]
        return grad_output * (x > 0).float()

def fn(x):
    return MyReLU.apply(x).sum()

# Works in both eager and compiled
fn(torch.randn(3, 3, requires_grad=True)).backward()
# [my_relu][fwd]
#   shape=torch.Size([3, 3]), norm=2.4495
# [my_relu][bwd]
#   shape=torch.Size([3, 3]), norm=3.0000
```

## Test plan

- `python test/higher_order_ops/test_opaque_inspect_tensor.py` — tests covering:
  - Single/multiple tensor inspection (eager, compile, make_fx, aot_function)
  - Forward output and gradient correctness
  - Graph structure verification (`assertExpectedInline` on fw/bw graphs)
  - Tag printing and multiple tags
  - `phase` parameter with custom autograd functions
  - Distributed rank filtering
  - `print(tensor)` repr compatibility

cc @mlazos

### CODE REVIEW

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 1]
```diff
@@ -367,6 +370,13 @@ def __call__(self, real_fn_spec, fake_fn_spec, *flat_args):
 
 invoke_leaf_function = InvokeLeafFunction()
 
+# Leaf functions are opaque and may have side effects (logging, printing, etc.).
+# Mark as side-effectful to prevent dead code elimination when outputs are unused.
+from torch.fx.node import has_side_effect
```
Comment (yushangdi) [inline torch/_higher_order_ops/invoke_leaf_function.py]: should remove this after https://github.com/pytorch/pytorch/pull/174122 lands

File: torch/_library/utils.py [hunk 2]
```diff
@@ -631,6 +631,11 @@ def is_impure(
             if args and len(args) > 0:
                 return args[0] in _side_effectful_functions
 
+        # TODO: should remove this after the effect_token PR lands
+        # https://github.com/pytorch/pytorch/pull/174122
```
Comment (yushangdi) [inline torch/_library/utils.py]: should remove this after #174122 lands

**Review by pianpwk — COMMENTED** [review]
maybe a dumb question, but suppose I want to use this on submodule boundaries, to log both activations and gradients, and contain FQN info, is there a way to do this? Maybe something like:
```python
  from torch.utils._debug_mode import DebugMode, hash_tensor_fn

  def install_debug_hooks(model):
      for name, module in model.named_modules():
          tag = f"{module.__class__.__name__}:{name}"

          def fwd_fn(t):
              mode = get_active_debug_mode()
              if mode is not None:
                  mode._handle_annotate(
                      f"[compile_print][fwd][{tag}] {t.shape} {t.dtype}: {hash_tensor_fn(t)}"
                  )

          def bwd_fn(grad):
              mode = get_active_debug_mode()
              if mode is not None:
                  mode._handle_annotate(
                      f"[compile_print][bwd][{tag}] {grad.shape} {grad.dtype}: {hash_tensor_fn(grad)}"
                  )

          print_fn = make_compile_print(fwd_fn, bwd_fn)

          def hook(mod, inp, output):
              pytree.tree_map_only(torch.Tensor, print_fn, output)

          module.register_forward_hook(hook)

install_debug_hooks(model)
```

This code is probably completely broken, but I'm wondering if there's a way to do it.

File: test/functorch/test_leaf_function.py [hunk 3]
```diff
@@ -372,7 +375,112 @@ def f(x, y):
...
+        )
+
+        with patch("builtins.print", lambda *args, **kwargs: printed.append(args)):
+            out_eager = f(x, y)
+            out_compiled = compiled_f(x_clone, y_clone)
+        self.assertEqual(out_eager, out_compiled)
+
+        # Both eager and compiled should have printed a tensor via print(z)
+        self.assertEqual(len(printed), 2)
+        self.assertIsInstance(printed[0][0], torch.Tensor)
```
Comment (fxdawnn) [inline test/functorch/test_leaf_function.py:511]: Can we also do ``assertEqual`` the printed[0][0] and the expected tensor value here? or are we not expected a specific output here with just the leaffunction?

File: test/higher_order_ops/test_compile_print.py [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+        y = torch.randn(3, 3, requires_grad=True)
+        buf = StringIO()
+        with redirect_stdout(buf):
+            out = compiled_f(x, y)
+            out.backward()
+        output = buf.getvalue()
+        self.assertIn("[A][fwd]", output)
+        self.assertIn("[B][fwd]", output)
+        self.assertIn("[A][bwd]", output)
+        self.assertIn("[B][bwd]", output)
```
Comment (fxdawnn) [inline test/higher_order_ops/test_compile_print.py]: Same for here too. I'm wondering if there is some reason to not test the actual tensor/grad being printed...

File: torch/_higher_order_ops/invoke_leaf_function.py (hunk 1 above)
Comment (fxdawnn) [reply torch/_higher_order_ops/invoke_leaf_function.py]: Did it pass the tests when you patch the PR locally?

**Review by fxdawnn — COMMENTED** [review]
Overall, it looks reasonable for using leaf_function to enable the fwd/bwd print to visualize the tensors. I do have some other comments for the testing pipeline.

File: torch/_higher_order_ops/compile_print_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
+from collections.abc import Callable
+
+import torch
+from torch._dynamo.decorators import leaf_function
+
+
+def make_compile_print(
```
Comment (SherlockNoMad) [inline torch/_higher_order_ops/compile_print_wrapper.py]: compile_print will also working eager, so I would suggest a more generic name. 

jax.debug.print is a good ref

File: torch/_higher_order_ops/compile_print_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        *tensors: torch.Tensor, tag: str = "", ranks: int | set[int] | None = None
+    ) -> tuple[torch.Tensor, ...]:
+        if _rank_enabled(ranks):
+            prefix = _make_prefix(tag)
+            if prefix:
+                print(f"{prefix}[fwd]")
+            for t in tensors:
+                if isinstance(t, torch.Tensor):
+                    fwd_f(t)
+        return (tensors[0].new_zeros(()),)
```
Comment (ydwu4) [inline torch/_higher_order_ops/compile_print_wrapper.py]: why we need return a tensor?

File: torch/_higher_order_ops/compile_print_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    return "".join(parts)
+
+
+def make_compile_print(
+    fwd_f: Callable[[torch.Tensor], None],
+    bwd_f: Callable[[torch.Tensor], None],
+) -> Callable[..., None]:
+    """
+    Create a callable that observes tensors during forward and backward passes,
+    compatible with eager mode, ``torch.compile``, and ``aot_function``.
```
Comment (ydwu4) [inline torch/_higher_order_ops/compile_print_wrapper.py]: for my understanding, what's "aot_function", is it a recommended user entrypoint?

Comment (ydwu4) [inline torch/_higher_order_ops/compile_print_wrapper.py]: We don't support inductor backend.

File: torch/_higher_order_ops/compile_print_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+          execute callbacks. ``None`` (default) means all ranks. An ``int`` means
+          only that rank. A ``set[int]`` means those ranks.
+
+    .. note::
+
+        ``bwd_f`` is called in both ``torch.compile`` and ``aot_function``, but
+        through different mechanisms:
+
+        - With ``torch.compile``, ``bwd_f`` runs as an autograd hook at runtime.
+          If ``cp`` is called on a **leaf input**, the hook fires *after* the
```
Comment (ydwu4) [inline torch/_higher_order_ops/compile_print_wrapper.py]: can explain what's "leaf input". don't want users to be confused with leaf function's input

File: torch/_higher_order_ops/compile_print_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+          compiled backward graph (so ``invoke_leaf_function`` does not appear in
```
Comment (ydwu4) [inline torch/_higher_order_ops/compile_print_wrapper.py]: Is there runtime behavior change when "invoke_leaf_function not appear in backward graph"?

File: test/higher_order_ops/test_compile_print.py [hunk 10]
```diff
@@ -0,0 +1,810 @@
...
+            fwd_f=lambda t: print("fwd", t),
+            bwd_f=lambda t: print("bwd", t),
+        )
+        buf = StringIO()
+        with redirect_stdout(buf):
+            run_eager(lambda _cp: make_fn(cp), inputs)
+        output = buf.getvalue()
+        self.assertIn("fwd", output)
+        self.assertIn("bwd", output)
+        # print(tensor) includes "tensor(" in its repr
```
Comment (ydwu4) [inline test/higher_order_ops/test_compile_print.py]: remove comment?

File: test/higher_order_ops/test_opaque_inspect_tensor.py [hunk 11]
```diff
@@ -0,0 +1,810 @@
...
+        cp2 = make_compile_print(
+            fwd_f=lambda t: fwd2.append(t.shape),
+            bwd_f=lambda t: bwd2.append(t.shape),
+        )
+        f = make_fn(cp1, cp2)
+        compiled_f = torch.compile(f, backend="aot_eager", fullgraph=True)
+        cloned = [x.clone().detach().requires_grad_(True) for x in inputs]
+        out = compiled_f(*cloned)
+        out.backward()
+        self.assertTrue(len(fwd1) >= 1)
```
Comment (ydwu4) [inline test/higher_order_ops/test_opaque_inspect_tensor.py:583]: can we use assertEqual for the the length of the lists?

File: torch/_higher_order_ops/compile_print_wrapper.py (hunk 7 above)
Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: it's just `make_fx` over `.grad()` call, so we get a joint graph. I think it's the current recommended entry point for getting a joint graph using non-strict approach, but it could change.

File: torch/_higher_order_ops/compile_print_wrapper.py (hunk 6 above)
Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: this was probably left over before None output is supported. removed

File: torch/_higher_order_ops/compile_print_wrapper.py (hunk 7 above)
Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: I changed the docstring to just say `make_fx`

Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: changed to ` ``torch.compile`` with ``aot_eager`` backend`

File: torch/_higher_order_ops/compile_print_wrapper.py (hunk 9 above)
Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: no, it should still print, just doesn't appear in the backward graph

Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: see `test_compile_graph_has_opaque_node` for an example of printing inputs. and see `test_compile_graph_has_opaque_node_in_bw` for an example  of printing intermediate tensor

File: torch/_higher_order_ops/compile_print_wrapper.py (hunk 8 above)
Comment (yushangdi) [reply torch/_higher_order_ops/compile_print_wrapper.py]: added some explanation!

**Review by ydwu4 — APPROVED** [review]
The implementation looks good to me. No strong opinions on the API though.
